tdigest/accumulation_tree.py

```diff
@@ -0,0 +1,151 @@
+import operator
+from bintrees.rbtree import RBTree, Node
+
+
+class AccumulationTree(RBTree):
+    def __init__(self, mapper, reducer=operator.add, zero=0):
+        RBTree.__init__(self)
+        self._zero = zero
+        self._mapper = mapper
+        self._reducer = reducer
+
+        self._dirty_nodes = set()
+
+        class AccumulatingNode(Node):
+            """Internal object, represents a tree node."""
+            __slots__ = ['accumulation']
+
+            def __init__(node, key, value):
+                Node.__init__(node, key, value)
+                node.accumulation = self._mapper(value)
+
+            def __setitem__(node, key, value):
+                Node.__setitem__(node, key, value)
+                self._dirty_nodes.add(node.key)
+
+        self._Node = AccumulatingNode
+
+
+    def _new_node(self, key, value):
+        """Create a new tree node."""
+        self._count += 1
+        node = self._Node(key, value)
+        return node
+
+    def insert(self, key, value):
+        try:
+            RBTree.insert(self, key, value)
+        finally:
+            self._update_dirty_nodes()
+
+    def remove(self, key):
+        try:
+            RBTree.remove(self, key)
+        finally:
+            self._update_dirty_nodes()
+
+    def _get_full_accumulation(self, node):
+        if not node:
+            return self._zero
+        else:
+            return node.accumulation
+
+    def _get_right_accumulation(self, node, lower):
+        if not node:
+            return self._zero
+        if(lower <= node.key):
+            return self._reducer(
+                self._reducer(
+                    self._get_right_accumulation(node[0], lower),
+                    self._get_full_accumulation(node[1]),
+                ),
+                self._mapper(node.value),
+            )
+        else:
+            return self._get_right_accumulation(node[1], lower)
+
+    def _get_left_accumulation(self, node, upper):
+        if not node:
+            return self._zero
+        if(upper > node.key):
+            return self._reducer(
+                self._reducer(
+                    self._get_full_accumulation(node[0]),
+                    self._get_left_accumulation(node[1], upper),
+                ),
+                self._mapper(node.value),
```

Comment:
Ah yes, I see where you are coming from. No, I don't think we're not adding the value twice. Because when we visit `node` in that function, we are not sure that we need the *entire* accumulation under `node`, which is why you'll notice `node.accumulation` is not accessed at all. This is the reason there is no double-counting: we *either* use `self._mapper(node.value)` *or* `node.accumulation`, but never both in the same function.

Let's go over it in detail. By definition, `_get_left_accumulation(self, node, upper)` returns the sum of all nodes `n` under `node` (including `node` itself) which satisfy `n.key < upper`. "By definition" is to say: that's what I want it to do. Let me try to convince you that indeed, it does:

```
def _get_left_accumulation(self, node, upper):
```
if node is None, this accumulation is zero. This is the stopping condition for the recursion.
```
        if not node:
            return self._zero
```
if `upper > node.key`, we now need to retrieve the sum of the following three things:
 - node.key itself (see definition);
 - the sum of `node.left` (`node[0]`) and all its children, because all of them have `n.key < node.key`, and therefore also `n.key < upper`. We call `_get_full_accumulation` on `node.left`;
 - that part of the tree rooted at `node.right` that satisfies `n.key < upper`. We just call this function recursively.
```
        if(upper > node.key):
            return self._reducer(
                self._reducer(
                    self._get_full_accumulation(node[0]),
                    self._get_left_accumulation(node[1], upper),
                ),
                self._mapper(node.value),
            )
```
if `upper <= node.key`, we can ignore this node and everything smaller, so we ignore `node[0]` and only recurse into `node[1]`.
```
        else:
            return self._get_left_accumulation(node[0], upper)
```
Does this explanation help? You can reason all other functions through in the same way.